### PR TITLE
Fix for cwstring, cconvert, transcode when !is_windows()

### PR DESCRIPTION
**Pull request by iagobaapellaniz**
On line `78` the function `cconvert()` calls for `transcode(Cwchar_t, String(s).data)` but latter it is not method for `trancode(::Type{UInt32}, Array{UInt8, 1})` on non windows machines.

I also chaged `Cwchar_t` from Int32 to UInt32, since I think it could be a typo.

This allows to pass `using PyCall` definitelly. See issue, [PyCall/#295](https://github.com/stevengj/PyCall.jl/issues/295).

More tests, with PyCall to come.

### CODE REVIEW

File: base/c.jl [hunk 1]
```diff
@@ -21,9 +21,9 @@ if is_windows()
     typealias Culong UInt32
     typealias Cwchar_t UInt16
 else
-    typealias Clong Int
```
Comment (i-apellaniz) [inline base/c.jl]: I changed from aliases to normal names, so it is clearer what's the difference between Windows and the rest of systems.

File: base/c.jl [hunk 2]
```diff
@@ -21,9 +21,9 @@ if is_windows()
...
-    typealias Culong UInt
-    typealias Cwchar_t Int32
+    typealias Clong Int64
+    typealias Culong UInt64
+    typealias Cwchar_t UInt32
```
Comment (i-apellaniz) [inline base/c.jl]: I changed from aliases to normal names, so it is clearer what's the difference between Windows and the rest of systems.

File: base/c.jl [hunk 3]
```diff
@@ -140,7 +140,7 @@ NULs are handled like any other character (i.e. the output will be NUL-terminate
 only if the `src` is).
 """
 function transcode end
-transcode{T<:Union{UInt8,UInt16}}(::Type{T}, src::Vector{T}) = src
+transcode{T<:Union{UInt8,UInt16,UInt32}}(::Type{T}, src::Vector{T}) = src
```
Comment (i-apellaniz) [inline base/c.jl]: This is not used for the fix, but I think it is reasonable for not Windows systems

File: base/c.jl [hunk 4]
```diff
@@ -21,9 +21,9 @@ if is_windows()
     typealias Culong UInt32
     typealias Cwchar_t UInt16
 else
-    typealias Clong Int
-    typealias Culong UInt
-    typealias Cwchar_t Int32
+    typealias Clong Int64
```
Comment (tkelman) [inline base/c.jl]: this is not correct on 32 bit unix

**Comment by i-apellaniz** [comment]
Should `Cwchar_t` be UInt32 or Int32 ??. I think it should be UInt32, as it is for Windows systems and as it is required by the methods called with this argument, for instance `transcode()`

File: base/c.jl [hunk 5]
```diff
@@ -23,7 +23,7 @@ if is_windows()
 else
     typealias Clong Int
     typealias Culong UInt
-    typealias Cwchar_t Int32
+    typealias Cwchar_t UInt32
```
Comment (yuyichao) [inline base/c.jl]: Why the change? `wchar_t` seems to be signed here.

Comment (i-apellaniz) [inline base/c.jl]: I know that Cwchar_t is passed as argument to the function `transcode` and it seems to accept on all the cases unsigned integers. So I changed the alias from `Int32` to `UInt32`. Also now it is coherent with the definition of the aliases for windows.

On the definition of `transcode{T::Union{UInt8, UInt16}(::Type{T}, src::T)=src`, line 143 and beyond.

I'm trying to build with appveyor but it is still running. I was able to build it locally with no error

Comment (yuyichao) [inline base/c.jl]: Consistency with windows is not the point for this definition. This is what the C type is.

Comment (yuyichao) [inline base/c.jl]: Also, changing this definition will be subtly breaking. Functions that need to use `Cwchar_t` should just be fixed to handle both cases.

Comment (i-apellaniz) [inline base/c.jl]: I'm not expertise on it, should I reverse it to `Int32`? Aren't the used to represent characters and therefore negatives are not useful?

Maybe `Int32` is faster than `UInt32` even if `UInt32` doesn't break anything.

Comment (yuyichao) [inline base/c.jl]: > should I reverse it to `Int32`

Yes.

> Aren't the used to represent characters and therefore negatives are not useful?

Blame whoever define them to be signed in C. `Cchar` is also signed on linux afaik.

File: base/c.jl [hunk 6]
```diff
@@ -143,8 +143,8 @@ function transcode end
 transcode{T<:Union{UInt8,UInt16}}(::Type{T}, src::Vector{T}) = src
 transcode(::Type{Int32}, src::Vector{UInt32}) = reinterpret(Int32, src)
 
-function transcode(::Type{UInt16}, src::Vector{UInt8})
-    dst = UInt16[]
+function transcode{T<:Union{UInt16, Int32}}(Dst_et::Type{T}, src::Vector{UInt8})
+    dst = Dst_et[]
```
Comment (yuyichao) [inline base/c.jl]: Why using `Dst_et` instead of `T` here?

File: base/c.jl [hunk 7]
```diff
@@ -143,8 +143,8 @@ function transcode end
...
     i, n = 1, length(src)
```
Comment (pabloferz) [inline base/c.jl]: I believe you can just write

```julia
function transcode{T<:Union{UInt16, Int32}}(::Type{T}, src::Vector{UInt8})
    dst = T[]
    ...
end
```

and use `T` instead of `Dst_et` which would look cleaner IMO.

File: base/c.jl [hunk 8]
```diff
@@ -143,8 +143,8 @@ function transcode end
 transcode{T<:Union{UInt8,UInt16}}(::Type{T}, src::Vector{T}) = src
 transcode(::Type{Int32}, src::Vector{UInt32}) = reinterpret(Int32, src)
 
-function transcode(::Type{UInt16}, src::Vector{UInt8})
-    dst = UInt16[]
+function transcode{T<:Union{UInt16, Int32}}(::Type{T}, src::Vector{UInt8})
```
Comment (vtjnash) [inline base/c.jl]: should be `T::Union{Type{UInt16}, Type{Int32}}`, without the static parameter

**Comment by i-apellaniz** [comment]
You all did the work instead me! Thank you. At least this fix let me use PyCall, and plot figures with PyPlot.

**Comment by stevengj** [comment]
The PyCall problem is fixed in PyCall `master`.

**Comment by stevengj** [comment]
`transcode` supports both `Int32` and `UInt32`, so defining `Cwchar_t` as `Int32` is not a problem for `transcode`.   It would seem cleaner if we defined it as `UInt32`, I agree.   The C standard [does not specify](http://stackoverflow.com/questions/11953363/wchar-t-is-unsigned-or-signed) whether `wchar_t` is signed.

On MacOS and Linux gcc, however, it looks like `wchar_t` is a (signed) `int` by default, so I think we were copying that.  (In principle, it probably wouldn't hurt if we used `UInt32` even if the C type is `Int32`, but for the most part we try to match signed-ness with the `Cxxxx` constants.)

**Comment by tkelman** [comment]
I would defer to / prefer being consistent with what GCC does on Linux, and what Clang does on Mac (if different than GCC).

**Comment by i-apellaniz** [comment]
I think that to be consistent, I should have to work on the reverse `transcode`

I've updated the code adding the `transcode(::Type{Int32}, src::Array{UInt8,1})`. I'll work on the reverse function for completeness.

**Comment by stevengj** [comment]
@i-apellaniz, I think you may be looking at an out-of-date version of Julia.  `Int32` support (in both directions) was already added to `transcode` in #17323.

**Comment by i-apellaniz** [comment]
Yes! It is added already. Let me testt it againts PyCall, and soon if everything is ok, i'll close this pull request.

Thanks all!
